cstar/base/additional_code.py

```diff
@@ -73,81 +88,139 @@ def __init__(
+                    checkout_target=self.checkout_target,
+                )
+                source_dir = Path(tmp_dir)
+            # CASE 2: Additional code is in a local directory/repository
+            elif (self.source.location_type == "path") and (
+                (self.source.source_type == "directory")
+                or (self.source.source_type == "repository")
+            ):
+                source_dir = Path(self.source.location)
+
+            else:
+                raise ValueError(
+                    "Invalid source for AdditionalCode. "
+                    + "AdditionalCode.source.location_type and "
+                    + "AdditionalCode.source.source_type should be "
+                    + "'url' and 'repository', or 'path' and 'repository', or"
+                    + "'path' and 'directory', not"
+                    + f"'{self.source.location_type}' and '{self.source.source_type}'"
+                )
+
+            # Now go through the files and copy them to local_dir
             for file_type in ["source_mods", "namelists"]:
                 file_list = getattr(self, file_type)
+
                 if file_list is None:
                     continue
-                tgt_dir = local_path + "/" + file_type + "/" + self.base_model.name
-                os.makedirs(tgt_dir, exist_ok=True)
+                tgt_dir = Path(local_dir) / file_type / self.base_model.name
+                tgt_dir.mkdir(parents=True, exist_ok=True)
+
                 for f in file_list:
-                    tmp_file_path = tmp_dir + "/" + f
-                    tgt_file_path = tgt_dir + "/" + os.path.basename(f)
-                    print("moving " + tmp_file_path + " to " + tgt_file_path)
-                    if os.path.exists(tmp_file_path):
-                        shutil.move(tmp_file_path, tgt_file_path)
+                    src_file_path = source_dir / f
+                    tgt_file_path = tgt_dir / Path(f).name
+                    print(f"copying {src_file_path} to {tgt_file_path}")
+                    if src_file_path.exists():
+                        shutil.copy(src_file_path, tgt_file_path)
                     else:
                         raise FileNotFoundError(
-                            f"Error: {tmp_file_path} does not exist."
+                            f"Error: {src_file_path} does not exist."
                         )
-        self.local_path = local_path
-        self.exists_locally = True
+                    # Special case for template namelists:
+                    if (
+                        file_type == "namelists"
+                        and str(src_file_path)[-9:] == "_TEMPLATE"
+                    ):
+                        print(
+                            f"copying {tgt_file_path} to editable namelist {str(tgt_file_path)[:-9]}"
+                        )
+                        shutil.copy(tgt_file_path, str(tgt_file_path)[:-9])
+                        if hasattr(self, "modified_namelists"):
```

Comment:
I think this needs an else to create the attribute if it's missing

Reply:
done in [a1a5bb6](https://github.com/CWorthy-ocean/C-Star/pull/55/commits/a1a5bb6bfef5c90986f060b629eb7faf0f6b8834)